io/src/main/scala/sbt/nio/Glob.scala

```diff
@@ -60,17 +61,284 @@ private[sbt] sealed trait ToGlob extends Any {
   def toGlob: Glob
 }
 object Glob extends LowPriorityGlobOps {
-  private[sbt] class ConvertedFileFilter(val f: FileFilter) extends PathNameFilter {
+  import scala.language.dynamics
+
+  /**
+   * Used to construct recursive globs using the `/` method of [[GlobOps]]
+   */
+  case object **
```

Comment:
Generally not a huge fan of symbolic ops (as I've mentioned about https://github.com/sbt/io/commit/e61821319b4e7c0074567d93e1428ad5a7a52aad adding English name for `def **` in 2015), but having two different meanings for `**` would be bad design.

Note that `*` (`glob`) and `**` (`globRecursive`) methods already return Glob in sbt 1.3:

```scala
scala> val g = base * "*.scala"
g: sbt.io.Glob =
Glob(
  base = /Users/eed3si9n/work/akka,
  filter = ConvertedFileFilter(ExtensionFilter(scala)),
  depth = 0
)

scala> val g = base ** "*.scala"
g: sbt.io.Glob =
Glob(
  base = /Users/eed3si9n/work/akka,
  filter = ConvertedFileFilter(ExtensionFilter(scala)),
  depth = 2147483647
)

scala> val g = base globRecursive "*.scala"
g: sbt.io.Glob =
Glob(
  base = /Users/eed3si9n/work/akka,
  filter = ConvertedFileFilter(ExtensionFilter(scala)),
  depth = 2147483647
)
```

Whatever the DSL we add here I think needs to be consistent with the rewrite we have to do in https://www.scala-sbt.org/1.x/docs/Paths.html. The case where a new DSL might be needed is expressing the new range.

I would propose adding an object named `anyPath` for this purpose, which can be used as:

```scala
scala> val g = base glob (anyPath / anyPath / "*.scala")
```

Reply:
I think the way to go is just to have `def /(string: String)` in GlobOps and use the https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileSystem.html#getPathMatcher(java.lang.String) so that `base / "**"`, `base / "*" / "*.txt"`, base / "*/*.txt" are all valid.

Reply:
Side note: the only reason that I proposed the `**` was because the `def /(**.type)` method only applied to `Path` and `Glob`. It could not be applied to `File`, so while `baseDirectory.value ** AllPassFilter` worked, `baseDirectory.value / **` did not. They did also return identical `Glob` instance so they were not actually significantly conceptually different. The point is moot though because I've already removed the `**` and `*` objects on my local brach. After I have lunch, I'm going to switch the glob parser to use the built in nio glob parser.

Reply:
Given `base` is `java.io.File`, would it require `toGlob` like:

```scala
base.toGlob / "*" / "*" / "*.scala"
```

?

Reply:
Yeah. Or base.toPath.

Reply:
Yeah. Or base.toPath.